### PR TITLE
Extend VHost API #8374

### PR DESCRIPTION
**Pull request by anatol-sialitski**

### CODE REVIEW

File: modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+exports.testGetVirtualHost = function () {
+    const result = libVhost.get('a');
+
+    t.assertEquals(JSON.stringify({
+        "name": "a",
+        "source": "/a",
+        "target": "/other/a",
+        "host": "localhost",
+        "defaultIdProviderKey": "default",
+        "idProviderKeys": [{"idProviderKey": "default"}]
```
Comment (jsi) [inline modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js:19]: Codacy found an issue: [Strings must use singlequote.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 12
           

Complexity increasing per file
==============================
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostsMapper.java  2
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostServiceImpl.java  3
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/VirtualHostHandler.java  1
- modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js  1
- modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js  2
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostMapper.java  3
- modules/lib/lib-vhost/src/test/java/com/enonic/xp/lib/vhost/VirtualHostHandlerTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598&bid=20253439)

File: modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    t.assertJsonEquals(true, result);
+};
+
+exports.testGetVirtualHost = function () {
+    const result = libVhost.get('a');
+
+    t.assertEquals(JSON.stringify({
+        "name": "a",
+        "source": "/a",
+        "target": "/other/a",
```
Comment (jsi) [inline modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js:16]: Codacy found an issue: [Strings must use singlequote.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * Functions to find virtual host.
+ *
+ * @example
+ * var vhostLib = require('/lib/xp/vhost');
+ *
+ * @module content
+ */
+
+function checkRequiredValue(value, name) {
+    if (value === undefined) {
```
Comment (jsi) [inline modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js:11]: Codacy found an issue: [Unexpected use of undefined.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+};
+
+/**
+ * This function returns virtual hosts.
+ *
+ * @returns {object} An object with all the virtual hosts.
+ */
+exports.list = function () {
+    const bean = __.newBean('com.enonic.xp.lib.vhost.VirtualHostHandler');
+    return __.toNativeObject(bean.getVirtualHosts());
```
Comment (jsi) [inline modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js:42]: Codacy found an issue: ['__' is not defined.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    t.assertJsonEquals(true, result);
+};
+
+exports.testGetVirtualHost = function () {
+    const result = libVhost.get('a');
+
+    t.assertEquals(JSON.stringify({
+        "name": "a",
+        "source": "/a",
```
Comment (jsi) [inline modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js:15]: Codacy found an issue: [Strings must use singlequote.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+function checkRequiredValue(value, name) {
+    if (value === undefined) {
+        throw 'Parameter \'' + name + '\' is required';
+    }
+}
+
+exports.isEnabled = function () {
+    const bean = __.newBean('com.enonic.xp.lib.vhost.VirtualHostHandler');
+    return __.toNativeObject(bean.isEnabled());
```
Comment (jsi) [inline modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js:18]: Codacy found an issue: ['__' is not defined.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * @param {string} name of the virtual host.
+ *
+ * @returns {object} The virtual host (as JSON).
+ */
+exports.get = function (name) {
+    checkRequiredValue(name, 'name');
+
+    const bean = __.newBean('com.enonic.xp.lib.vhost.VirtualHostHandler');
+    return __.toNativeObject(bean.getVirtualHost(name));
```
Comment (jsi) [inline modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js:32]: Codacy found an issue: ['__' is not defined.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        "defaultIdProviderKey": "default",
+        "idProviderKeys": [{"idProviderKey": "default"}]
+    }), JSON.stringify(result));
+};
+
+exports.testGetVirtualHosts = function () {
+    const result = libVhost.list();
+
+    t.assertEquals(JSON.stringify({
+        "vhosts": [{
```
Comment (jsi) [inline modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js:27]: Codacy found an issue: [Strings must use singlequote.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+exports.testGetVirtualHost = function () {
+    const result = libVhost.get('a');
+
+    t.assertEquals(JSON.stringify({
+        "name": "a",
+        "source": "/a",
+        "target": "/other/a",
+        "host": "localhost",
```
Comment (jsi) [inline modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js:17]: Codacy found an issue: [Strings must use singlequote.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+function checkRequiredValue(value, name) {
+    if (value === undefined) {
+        throw 'Parameter \'' + name + '\' is required';
+    }
+}
+
+exports.isEnabled = function () {
+    const bean = __.newBean('com.enonic.xp.lib.vhost.VirtualHostHandler');
```
Comment (jsi) [inline modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js:17]: Codacy found an issue: ['__' is not defined.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        "defaultIdProviderKey": "default",
```
Comment (jsi) [inline modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js:18]: Codacy found an issue: [Strings must use singlequote.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    const result = libVhost.isEnabled();
+
+    t.assertJsonEquals(true, result);
+};
+
+exports.testGetVirtualHost = function () {
+    const result = libVhost.get('a');
+
+    t.assertEquals(JSON.stringify({
+        "name": "a",
```
Comment (jsi) [inline modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js:14]: Codacy found an issue: [Strings must use singlequote.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

File: modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostService.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+public interface VirtualHostService
+{
+
+    boolean isEnabled();
+
+    VirtualHost getVirtualHost( final String name );
+
+    List<VirtualHost> getVirtualHosts();
+
+    VirtualHost resolve( final HttpServletRequest req );
```
Comment (rymsha) [inline modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostService.java:16]: isEnabled, getVirtualHost, getVirtualHosts should be in a different service (VirtualHostConfigService)
resolve should stay here.

**Review by rymsha — CHANGES_REQUESTED** [review]

File: modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostService.java (hunk 13 above)
Comment (anatol-sialitski) [reply modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostService.java]: done

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/VirtualHostHandler.java  1
- modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js  1
- modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostHelper.java  2
- modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js  2
- modules/lib/lib-vhost/src/test/java/com/enonic/xp/lib/vhost/VirtualHostHandlerTest.java  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostServiceImpl.java  3
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostMapper.java  3
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostsMapper.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598&bid=20253439)

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public boolean isEnabled()
+    {
+        return virtualHostConfig.isEnabled();
+    }
+
+    @Override
+    public VirtualHost getVirtualHost( final String name )
```
Comment (rymsha) [inline modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostServiceImpl.java:33]: I now hardly see a need for such an API method.
1. In java one can easily use streams
2- in js one can use find

**Review by rymsha — CHANGES_REQUESTED** [review]

File: modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostHelper.java [hunk 15]
```diff
@@ -23,4 +29,57 @@ public static void setVirtualHost( final HttpServletRequest req, final VirtualHo
     {
         req.setAttribute( KEY, vhost );
     }
+
+    public static VirtualHost resolve( final List<VirtualHost> virtualHosts, final HttpServletRequest req )
```
Comment (rymsha) [inline modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostHelper.java:33]: Move to own service VirtualHostResolverService

File: modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostHelper.java [hunk 16]
```diff
@@ -23,4 +29,57 @@ public static void setVirtualHost( final HttpServletRequest req, final VirtualHo
...
+            if ( matches( virtualHost, req ) )
+            {
+                return virtualHost;
+            }
+        }
+
+        return null;
+    }
+
+    public static boolean matches( final VirtualHost virtualHost, final HttpServletRequest req )
```
Comment (rymsha) [inline modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostHelper.java:46]: move to another helper (internal package)

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostConfig.java [hunk 17]
```diff
@@ -1,10 +1,12 @@
 package com.enonic.xp.web.vhost.impl.config;
 
-import com.enonic.xp.web.vhost.impl.mapping.VirtualHostMappings;
+import java.util.List;
+
+import com.enonic.xp.web.vhost.VirtualHost;
 
 public interface VirtualHostConfig
```
Comment (rymsha) [inline modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostConfig.java:7]: This service is not needed. Replace by VirtualHostService(Impl)

**Review by rymsha — CHANGES_REQUESTED** [review]

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/lib/lib-vhost/src/test/java/com/enonic/xp/lib/vhost/VirtualHostHandlerTest.java  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImpl.java  4
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostMapper.java  3
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java  2
- modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js  1
- modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostsMapper.java  2
- modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostHelper.java  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/VirtualHostHandler.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598&bid=20253439)

File: modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostHelper.java [hunk 18]
```diff
@@ -23,4 +27,16 @@ public static void setVirtualHost( final HttpServletRequest req, final VirtualHo
     {
         req.setAttribute( KEY, vhost );
     }
+
+    public static String normalizePath( final String value )
```
Comment (rymsha) [inline modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostHelper.java:31]: Hopefully we can get rid of this method.

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImpl.java [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    private final VirtualHostService virtualHostService;
+
+    @Activate
+    public VirtualHostResolverImpl( @Reference final VirtualHostService virtualHostService )
+    {
+        this.virtualHostService = virtualHostService;
+    }
+
+    @Override
+    public VirtualHost resolveVirtualHost( final HttpServletRequest req )
```
Comment (rymsha) [inline modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImpl.java:28]: XP has WebRequest interface abstraction. maybe that should be used instead?

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostConfigMap.java [hunk 20]
```diff
@@ -34,14 +37,19 @@ public boolean isEnabled()
...
-        final VirtualHostMappings mappings = new VirtualHostMappings();
+        final List<VirtualHost> mappings = new ArrayList<>();
+
         for ( final String name : findMappingNames() )
         {
             mappings.add( buildMapping( name ) );
         }
 
+        mappings.sort( Comparator.comparing( VirtualHost::getHost ).
+            thenComparing( VirtualHost::getSource, ( virtualHost, otherVirtualHost ) -> otherVirtualHost.length() - virtualHost.length() ).
```
Comment (rymsha) [inline modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostConfigMap.java:50]: themComparingInt

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostConfigMap.java [hunk 21]
```diff
@@ -34,14 +37,19 @@ public boolean isEnabled()
...
+            thenComparing( VirtualHost::getSource, String::compareTo ) );
```
Comment (rymsha) [inline modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostConfigMap.java:51]: `String::compareTo` is it needed? I think method is overloaded.

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostConfigMap.java [hunk 22]
```diff
@@ -34,14 +37,19 @@ public boolean isEnabled()
         return getBoolean( "enabled", false );
     }
 
-    public VirtualHostMappings buildMappings()
+    public List<VirtualHost> buildMappings()
     {
-        final VirtualHostMappings mappings = new VirtualHostMappings();
+        final List<VirtualHost> mappings = new ArrayList<>();
+
         for ( final String name : findMappingNames() )
```
Comment (rymsha) [inline modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostConfigMap.java:44]: `findMappingNames().stream().map(...).sort().collect()`

**Review by rymsha — CHANGES_REQUESTED** [review]

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostMapper.java  3
- modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostHelper.java  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/VirtualHostHandler.java  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImpl.java  4
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java  2
- modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostsMapper.java  2
- modules/lib/lib-vhost/src/test/java/com/enonic/xp/lib/vhost/VirtualHostHandlerTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598&bid=20253439)

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java  2
- modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/VirtualHostHandler.java  1
- modules/lib/lib-vhost/src/test/java/com/enonic/xp/lib/vhost/VirtualHostHandlerTest.java  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImpl.java  4
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostInternalHelper.java  2
- modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostMapper.java  3
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostsMapper.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598&bid=20253439)

**Comment by rymsha** [comment]
DO NOT MERGE before 7.5 release

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java [hunk 23]
```diff
@@ -1,28 +1,31 @@
...
 
-    private VirtualHostMappings mappings;
+    private List<VirtualHost> virtualHosts;
 
-    public VirtualHostConfigImpl()
+    public VirtualHostServiceImpl()
     {
         this.enabled = false;
-        this.mappings = new VirtualHostMappings();
+        this.virtualHosts = new ArrayList<>();
```
Comment (rymsha) [inline modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java:28]: List.of()

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java [hunk 24]
```diff
@@ -32,17 +35,17 @@ public boolean isEnabled()
...
+        return this.virtualHosts;
     }
 
     @Activate
     public void configure( final Map<String, String> config )
     {
         final VirtualHostConfigMap configMap = new VirtualHostConfigMap( config );
         this.enabled = configMap.isEnabled();
-        this.mappings = configMap.buildMappings();
+        this.virtualHosts = configMap.buildMappings();
```
Comment (rymsha) [inline modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java:48]: List.copyOf(configMap.buildMappings())

**Review by rymsha — CHANGES_REQUESTED** [review]

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java [hunk 25]
```diff
@@ -1,28 +1,31 @@
 package com.enonic.xp.web.vhost.impl.config;
 
+import java.util.ArrayList;
```
Comment (jsi) [inline modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java:3]: Codacy found an issue: [Unused import - java.util.ArrayList.](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598)

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js  1
- modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java  2
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/VirtualHostHandler.java  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostsMapper.java  2
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostInternalHelper.java  2
- modules/lib/lib-vhost/src/test/java/com/enonic/xp/lib/vhost/VirtualHostHandlerTest.java  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImpl.java  4
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostMapper.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598&bid=20253439)

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java (hunk 23 above)
Comment (anatol-sialitski) [reply modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java]: done

File: modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java (hunk 24 above)
Comment (anatol-sialitski) [reply modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java]: done

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImpl.java  4
- modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java  2
- modules/lib/lib-vhost/src/test/java/com/enonic/xp/lib/vhost/VirtualHostHandlerTest.java  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostsMapper.java  2
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/VirtualHostHandler.java  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostMapper.java  3
- modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostInternalHelper.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598&bid=20253439)

File: modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostResolver.java [hunk 26]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.enonic.xp.annotation.PublicApi;
+
+@PublicApi
+public interface VirtualHostResolver
+{
+
+    VirtualHost resolveVirtualHost( final HttpServletRequest req );
+
+    String getFullTargetPath( final VirtualHost virtualHost, final HttpServletRequest req );
```
Comment (rymsha) [inline modules/web/web-api/src/main/java/com/enonic/xp/web/vhost/VirtualHostResolver.java:13]: move to VirtualHostInternalHelper

**Review by rymsha — CHANGES_REQUESTED** [review]

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java  2
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostsMapper.java  2
- modules/lib/lib-vhost/src/test/java/com/enonic/xp/lib/vhost/VirtualHostHandlerTest.java  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImpl.java  4
- modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/VirtualHostHandler.java  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostInternalHelper.java  2
- modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostMapper.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598&bid=20253439)

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImpl.java  4
- modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostMapper.java  3
- modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/VirtualHostHandler.java  1
- modules/lib/lib-vhost/src/test/java/com/enonic/xp/lib/vhost/VirtualHostHandlerTest.java  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java  2
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostsMapper.java  2
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostInternalHelper.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=6210598&bid=20253439)

**Comment by rymsha** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/lib/lib-vhost/src/test/java/com/enonic/xp/lib/vhost/VirtualHostHandlerTest.java  1
- modules/web/web-vhost/src/test/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImplTest.java  1
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostsMapper.java  2
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/mapper/VirtualHostMapper.java  3
- modules/web/web-vhost/src/test/java/com/enonic/xp/web/vhost/impl/VirtualHostInternalHelperTest.java  1
- modules/lib/lib-vhost/src/main/resources/lib/xp/vhost.js  1
- modules/lib/lib-vhost/src/test/resources/com/enonic/xp/lib/vhost/vhost-test.js  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostInternalHelper.java  3
- modules/lib/lib-vhost/src/main/java/com/enonic/xp/lib/vhost/VirtualHostHandler.java  1
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/VirtualHostResolverImpl.java  4
- modules/web/web-vhost/src/test/java/com/enonic/xp/web/vhost/impl/VirtualHostServiceImplTest.java  3
- modules/web/web-vhost/src/main/java/com/enonic/xp/web/vhost/impl/config/VirtualHostServiceImpl.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=6220525&bid=20277515)

**Review by rymsha — APPROVED** [review]

**Review by vbradnitski — APPROVED** [review]

**Pull request merged**